feat(nix): add nix flake with dev shell

doprz opened the pull request:
- **PR Description**

This PR adds Nix flake support to make contributing to lazygit easier for developers as it creates a reproducible development environment including go and other deps.

- **Please check if the PR fulfills these requirements**

n/a as this PR doesn't modify go files.

* [ ] Cheatsheets are up-to-date (run `go generate ./...`)
* [ ] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [ ] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [ ] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [ ] If a new UserConfig entry was added, make sure it can be hot-reloaded (see [here](https://github.com/jesseduffield/lazygit/blob/master/docs/dev/Codebase_Guide.md#using-userconfig))
* [ ] Docs have been updated if necessary
* [ ] You've read through your own file changes for silly mistakes etc

<!--
Be sure to name your PR with an imperative e.g. 'Add worktrees view', and make sure the title
is suitable to be included as a bullet point in release notes (i.e. phrased from a user's point
of view).
see https://github.com/jesseduffield/lazygit/releases/tag/v0.40.0 for examples
-->

Eveeifyeve commented on .gitignore:
```diff
@@ -28,3 +28,10 @@ __debug_bin*
+result-*
+
+# Ignore automatically generated direnv output
+.direnv
```
Please add .envrc to ignore please...

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,126 @@
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-25.05";
+    flake-utils.url = "github:numtide/flake-utils";
```
Please use flake-parts instead as flake-utils doesn't follow the module system.

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,126 @@
+  outputs = { self, nixpkgs, flake-utils }:
+    let
+      supportedSystems =
+        [ "x86_64-linux" "aarch64-linux" "x86_64-darwin" "aarch64-darwin" ];
```
Use nix-systems please.

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,126 @@
+{
+  description = "A simple terminal UI for git commands";
+
+  inputs = {
```
Add flake-compact for compatibility for users or move the packages and shells in separate files who don't use flakes.

Eveeifyeve commented on README.md:
````diff
@@ -374,19 +374,70 @@ sudo zypper ar https://download.opensuse.org/repositories/devel:/languages:/go/$
+
+**Installation with flakes:**
+```sh
+nix profile install github:jesseduffield/lazygit
````
I highly suggest not recommend the profile install command as it is not the nix way of doing.

Eveeifyeve reviewed (changes requested):
I like what you are doing, but a lot of changes needed.

stefanhaller commented:
I really wonder if this thing should be maintained in our repo at all. It sounds similar to an rpm spec file, a deb file, or a homebrew formula, and we don't maintain those here either; we let people who are packaging experts do it wherever it makes sense to them. I'd feel better about this if we took the same approach for this flake thing too. (Happy to link to it from the readme though, of course.)

Or am I missing something about why it conceptionally makes sense to maintain this here, unlike those other ones?

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,124 @@
+            inherit lazygit;
+          };
+
+          apps = {
+            default = {
+              type = "app";
+              program = "${lazygit}/bin/lazygit";
+            };
+            lazygit = {
+              type = "app";
+              program = "${lazygit}/bin/lazygit";
+            };
+          };
```
Why define apps when you have packages?

doprz commented on flake.nix:
```diff
@@ -0,0 +1,124 @@
+        };
+
+        # CI/CD support
+        hydraJobs = inputs.self.packages;
```
I had this as I was using Hydra (NixOS's CI/CD system) and forgot to remove it. Good catch.

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,110 @@
+            CGO_ENABLED = "0";
+          };
+
+          formatter = pkgs.nixfmt-rfc-style;
```
Consider treefmt-nix not needed.

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,110 @@
+
+          lazygit = pkgs.buildGoModule rec {
+            pname = "lazygit";
+            inherit version;
+
+            src = ./.;
+            vendorHash = null;
+
+            # Disable integration tests that require specific environment
+            doCheck = false;
+
+            nativeBuildInputs = with pkgs; [
+              git
+              makeWrapper
+            ];
+            buildInputs = with pkgs; [ git ];
+
+            ldflags = [
+              "-s"
+              "-w"
+              "-X main.commit=${gitCommit}"
+              "-X main.version=${version}"
+              "-X main.buildSource=nix"
+            ];
+
+            postInstall = ''
+              wrapProgram $out/bin/lazygit \
+                --prefix PATH : ${pkgs.lib.makeBinPath [ pkgs.git ]}
+            '';
+
+            meta = with pkgs.lib; {
+              description = "A simple terminal UI for git commands";
+              homepage = "https://github.com/jesseduffield/lazygit";
+              license = licenses.mit;
+              maintainers = [ "jesseduffield" ];
+              platforms = platforms.unix;
+              mainProgram = "lazygit";
+            };
+          };
+        in
```
Why not move version and gitCommit inside of lazygit?

doprz replied:
`buildGoModule` requires the version attribute to be set or else it won't build. I tried removing both the `gitCommit` and `version` attribute seeing as I found this section of code under `pkgs/app/entry_point.go` however `commit` isn't generated properly: `commit=, build date=, build source=nix, version=0.54.2, os=linux, arch=amd64, git version=2.50.1`

 It works as expected when running `make build`.

```go
// binary via the LDFLAGS argument. Some lazygit distributions will make use of these
// arguments and some will not. Go recently started baking in build info
// into the binary by default e.g. the git commit hash. So in this function
// we merge the two together, giving priority to the stuff set by LDFLAGS.
// Note: this mutates the argument passed in
func mergeBuildInfo(buildInfo *BuildInfo) {
	// if the version has already been set by build flags then we'll honour that.
	// chances are it's something like v0.31.0 which is more informative than a
	// commit hash.
	if buildInfo.Version != "" {
		return
	}

	buildInfo.Version = "unversioned"

	goBuildInfo, ok := debug.ReadBuildInfo()
	if !ok {
		return
	}

	revision, ok := lo.Find(goBuildInfo.Settings, func(setting debug.BuildSetting) bool {
		return setting.Key == "vcs.revision"
	})
	if ok {
		buildInfo.Commit = revision.Value
		// if lazygit was built from source we'll show the version as the
		// abbreviated commit hash
		buildInfo.Version = utils.ShortHash(revision.Value)
	}

	// if version hasn't been set we assume that neither has the date
	time, ok := lo.Find(goBuildInfo.Settings, func(setting debug.BuildSetting) bool {
		return setting.Key == "vcs.time"
	})
	if ok {
		buildInfo.Date = time.Value
	}
}
```

doprz replied:
Open to suggestions on this.

Eveeifyeve replied:
But I would suggest move version inside of it and then use the powerful rec.

doprz replied:
Fixed

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,110 @@
+              gnumake
+            ];
+
+            shellHook = ''
+              echo "Lazygit development environment"
+              echo "Go version: $(go version)"
+              echo "Git version: $(git --version)"
+              echo ""
+            '';
```
Not needed, Please remove.

Eveeifyeve reviewed (changes requested):
Only two more changes needed. Before I consider this LGTM.

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,112 @@
+
+          lazygit = pkgs.buildGoModule rec {
+            pname = "lazygit";
+            inherit version;
+
+            src = ./.;
+            vendorHash = null;
+
+            # Disable integration tests that require specific environment
+            doCheck = false;
+
+            nativeBuildInputs = with pkgs; [
+              git
+              makeWrapper
+            ];
+            buildInputs = with pkgs; [ git ];
+
+            ldflags = [
+              "-s"
+              "-w"
+              "-X main.commit=${gitCommit}"
+              "-X main.version=${version}"
+              "-X main.buildSource=nix"
+            ];
+
+            postInstall = ''
+              wrapProgram $out/bin/lazygit \
+                --prefix PATH : ${pkgs.lib.makeBinPath [ pkgs.git ]}
+            '';
+
+            meta = with pkgs.lib; {
+              description = "A simple terminal UI for git commands";
+              homepage = "https://github.com/jesseduffield/lazygit";
+              license = licenses.mit;
+              maintainers = [ "jesseduffield" ];
+              platforms = platforms.unix;
+              mainProgram = "lazygit";
+            };
+          };
+        in
```
```suggestion
          lazygit = pkgs.buildGoModule rec {
            pname = "lazygit";
            version = "0.54.2";
            
            gitCommit = inputs.self.rev or inputs.self.dirtyRev or "dev";

            src = ./.;
            vendorHash = null;

            # Disable integration tests that require specific environment
            doCheck = false;

            nativeBuildInputs = with pkgs; [
              git
              makeWrapper
            ];
            buildInputs = with pkgs; [ git ];

            ldflags = [
              "-s"
              "-w"
              "-X main.commit=${gitCommit}"
              "-X main.version=${version}"
              "-X main.buildSource=nix"
            ];

            postInstall = ''
              wrapProgram $out/bin/lazygit \
                --prefix PATH : ${pkgs.lib.makeBinPath [ pkgs.git ]}
            '';

            meta = with pkgs.lib; {
              description = "A simple terminal UI for git commands";
              homepage = "https://github.com/jesseduffield/lazygit";
              license = licenses.mit;
              maintainers = [ "jesseduffield" ];
              platforms = platforms.unix;
              mainProgram = "lazygit";
            };
          };
        in
```

doprz commented:
Thank you for your feedback @Eveeifyeve on this PR! It's been very helpful to learn best practices.

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,111 @@
+            programs.gofmt.enable = true;
+          };
+
+          checks = {
+            # Ensure the package builds
+            build = lazygit;
+          };
```
```suggestion
          checks.build = lazygit;
```

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,111 @@
+  outputs =
+    inputs@{ flake-parts, systems, ... }:
+    flake-parts.lib.mkFlake { inherit inputs; } {
+      imports = [
+        inputs.treefmt-nix.flakeModule
+      ];
+
+      systems = import systems;
```
```suggestion
      systems = import systems;
      imports = [
        inputs.treefmt-nix.flakeModule
      ];
```

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,111 @@
+              git
+              makeWrapper
+            ];
+            buildInputs = with pkgs; [ git ];
```
```suggestion
            buildInputs = [ pkgs.git ];
```

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,111 @@
+                --prefix PATH : ${pkgs.lib.makeBinPath [ pkgs.git ]}
+            '';
+
+            meta = with pkgs.lib; {
+              description = "A simple terminal UI for git commands";
+              homepage = "https://github.com/jesseduffield/lazygit";
+              license = licenses.mit;
+              maintainers = [ "jesseduffield" ];
+              platforms = platforms.unix;
+              mainProgram = "lazygit";
+            };
```
```suggestion
            meta = {
              description = "A simple terminal UI for git commands";
              homepage = "https://github.com/jesseduffield/lazygit";
              license = pkgs.lib.licenses.mit;
              maintainers = [ "jesseduffield" ];
              platforms = pkgs.lib.platforms.unix;
              mainProgram = "lazygit";
            };
```

Read: https://github.com/NixOS/nixpkgs/issues/208242

Eveeifyeve reviewed (approved):
A few nit picks, real issue is the `with pkgs.lib` else it's LGTM.

doprz commented:
Thank you @Eveeifyeve!
cc: @stefanhaller This PR is now complete.

Eveeifyeve reviewed (approved):
![](https://raw.githubusercontent.com/onevcat/lgtm-images/master/images/VPgV6sn.png)

doprz commented:
@stefanhaller bump

stefanhaller reviewed (commented):
A few things.

- The branch is totally unreviewable with all the merges from master, and all the fine-grained commits that change things back and forth. I rebased onto master to get rid of the merge commits and force-pushed, this makes it a little bit reviewable at least. In the end I suppose all commits need to be squashed into one. We don't do this by default here, it's the contributor's job to provide a commit sequence that makes sense. In the future, please look into working with `fixup!` commits, and never merge master into your branch; always rebase onto master instead.
- As for maintaining it here vs. elsewhere: your explanation above still didn't make it very clear to me what the benefit is of maintaining the flake in this repo vs. somewhere else. I now saw that we have an issue for this (#3474), this explains it a little bit better. But still, there are also things like flakehub, would it be an option to use that instead?

The reason why I'm still hesitant to host the flake here is that this PR has shown that it needs expertise to get this right. @Eveeifyeve seems to have done a great job at helping with this, but if they are not around next time we need to make a change to this, who else will be able to do that? I won't, because I totally lack the expertise, and I don't intend to acquire it. @Eveeifyeve What's your take on this?

Eveeifyeve commented on CONTRIBUTING.md:
````diff
@@ -60,6 +60,31 @@ To run lazygit from within the integrated terminal just go `go run main.go`
+```
+
+This will drop you into a development shell that includes:
+* Go 1.24 toolchain
````
Totally agree with this suggestion.

```suggestion
* Go toolchain
```

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,107 @@
+
+            buildInputs = with pkgs; [
+              # Go toolchain
+              go_1_24
```
I would say base it on the latest go version.

Here is a suggestion to fix this:
```suggestion
              go
```

stefanhaller replied:
I don't think this is appropriate for guaranteeing a stable environment, which I understand is the whole point here.

While go promises not to break the build with language updates, it might still break tests. It looks like this happened in #4844, so currently developing lazygit works with 1.24 but not with 1.25. (I actually haven't investigated those test failures yet, so I don't even know if these only affect running the tests or also introduce regressions in the production code.)

Eveeifyeve replied:
Well it's impossible of getting the version from the go mod unless you do a script with `sed` and some magic like make it a package in list which I can't possibly do myself or require huge amounts of time to make it work and many go projects add a defaulted go version to use. If you want to try go right ahead, but for now we should stick to the latest go package since we can't read the go mod file.

stefanhaller replied:
@doprz Please don't mark things as resolved when they aren't.

@Eveeifyeve In that case we should go back to specifying the version explicitly and accept the redundancy. We do the same in the Dockerfile and in `.github/workflows/ci.yml` and other places. It is important to be on the same version everywhere to guarantee predictable results.

The problem is that those other places can easily be found by a global search, but the one in the flake can't because it uses a different format. To mitigate this, we should probably add a comment such as:
```diff
diff --git i/go.mod w/go.mod
index 324be6d7e..f48ce401b 100644
--- i/go.mod
+++ w/go.mod
@@ -1,5 +1,7 @@
 module github.com/jesseduffield/lazygit
 
+// If you change this, be sure to also update flake.nix and other files that
+// mention the go version (search for it)
 go 1.24.0
 
 require (
```

doprz replied:
This can be solved via:

```nix
  goMod = builtins.readFile ./go.mod;
  goVersion = builtins.head (builtins.match ".*go ([0-9]+\.[0-9]+).*" goMod);
  go = pkgs."go_${builtins.replaceStrings ["."] ["_"] goVersion}";
```

<img width="678" height="315" alt="image" src="https://github.com/user-attachments/assets/22417659-d68f-47f5-9282-b165e2a9147c" />

However this doesn't follow the Nix Philosophy. Thoughts @Eveeifyeve?

stefanhaller replied:
@Eveeifyeve Good point, I guess the regex could be a bit stricter, maybe something like `"^go\\s+(\\d+\\.\\d+)\\.\\d+$"`. We don't need to be lenient against the version having either two or three components; we know it will always be three.

@doprz Did you actually test your version? I think you need to double the `\` in the pattern to escape it properly.

Can we add some error checking to make sure the whole thing errors out when it doesn't find a match?

Finally, I'd consider matching the first two components of the version separately so that you don't need to replace dots by underscores afterwards.

doprz replied:
> Yeah that could work, but only problem isn't this searching all go's in the file which could cause a problem right?

True but then this becomes into an issue of making sure that all go version instances are matching. I would suppose that the maintainers make sure that their go version throughout the repo matches and if that's the case, then the go version that the nix flake pulls will be consistent, however we could add a section to the makefile that checks all instances of go versions as a check.

doprz replied:
Nix doesn't support the `\s` and `\d` regex shortcuts so I've adjusted the regex to work properly and error when required.

stefanhaller commented:
> it should be in the official repo because packages take a few days before they reach unstable 

That alone is not a reason to host them here though, is it? The same can be said about any distribution or package manager. Look at [this table](https://github.com/jesseduffield/lazygit?tab=readme-ov-file#installation), where most table cells need a few days to become green after a new release. We are not going to start hosting deb files, rpm spec files, or homebrew formula files in this repo to avoid that.

doprz commented:
This configuration is at the point where it should not require any major changes or foreseeable maintenance to it. The main purpose of this PR was to allow nix users to have a development shell with all of the required deps and tools to start developing and using `lazygit`. I've also added the ability to build with nix as a nice bonus. I've also squashed the commits + updated the docs as requested.

The original commit I made worked albeit not using best practices as I wasn't aware of things such as `flake-compat` to offer backwards compatibility to non-flake users. A good example of a project offering nix configuration with a dev environment is [ghostty](https://github.com/ghostty-org/ghostty). I'm easily able to clone the repository and have a working zig environment that is reproducible.

<img width="515" height="156" alt="image" src="https://github.com/user-attachments/assets/670e2313-607e-4cff-9ee1-4d012b099bc7" />

The PR is not about replacing the "unofficial" nix flake or adding distribution/package management files to the lazygit repo. It's about allowing nix users to have a declarative and reproducible dev environment to allow them to contribute to lazygit. All they have to do is cd into the repo and load up the dev shell or build lazygit via nix or with make. It's a quality of life PR at it's core.

stefanhaller commented on flake.nix:
```diff
@@ -0,0 +1,107 @@
+              # Go toolchain
+              go
+              gotools
+              golangci-lint
```
For golangci-lint it is even more important than for go itself to be on the right version. Which checks are enabled by default tends to change between versions, so you are likely to get spurious linter warnings when using the latest version.

We already have a mechanism in place to mitigate this by using a shim that runs golangci-lint with the right version (see `make lint` in the Makefile), so I wonder if we should remove this entirely rather than add an explicit version.

doprz replied:
Got it. Yes, then we can remove this from the nix file as I'm including gnumake. Will wait for unresolved comments to be updated before pushing a chore commit and rebasing.

Eveeifyeve commented:
> > This configuration is at the point where it should not require any major changes or foreseeable maintenance to it. The main purpose of this PR was to allow nix users to have a development shell with all of the required deps and tools to start developing and using `lazygit`. I've also added the ability to build with nix as a nice bonus. I've also squashed the commits + updated the docs as requested.
> > The original commit I made worked albeit not using best practices as I wasn't aware of things such as `flake-compat` to offer backwards compatibility to non-flake users. A good example of a project offering nix configuration with a dev environment is [ghostty](https://github.com/ghostty-org/ghostty?rgh-link-date=2025-09-19T08%3A20%3A10Z). I'm easily able to clone the repository and have a working zig environment that is reproducible.
> > <img alt="image" width="515" height="156" src="https://private-user-images.githubusercontent.com/52579214/486186118-670e2313-607e-4cff-9ee1-4d012b099bc7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgyNjk1NzEsIm5iZiI6MTc1ODI2OTI3MSwicGF0aCI6Ii81MjU3OTIxNC80ODYxODYxMTgtNjcwZTIzMTMtNjA3ZS00Y2ZmLTllZTEtNGQwMTJiMDk5YmM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTE5VDA4MDc1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM0MTdkNjUzMWZhZTYwMWNmY2I4OTI3ZTVjNWE1ZDIzZGI2N2IzZTdhNjcxNDdhYjI5ODE5MzQ5ZmZkOTRiYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pBk1Tzv8J1Tc2WHPg3Xeags_HiwJ8gWbfUCmy2LIvEw">
> > The PR is not about replacing the "unofficial" nix flake or adding distribution/package management files to the lazygit repo. It's about allowing nix users to have a declarative and reproducible dev environment to allow them to contribute to lazygit. All they have to do is cd into the repo and load up the dev shell or build lazygit via nix or with make. It's a quality of life PR at it's core.
> 
> so its like a [dev container](https://containers.dev/)? but for nix users? am i understanding this correctly?

please just take a look at https://nixos.org/learn/

stefanhaller commented:
@doprz Are you waiting for any more input? I thought there's consensus about the way forward for the two open issues.

doprz commented:
> @doprz Are you waiting for any more input? I thought there's consensus about the way forward for the two open issues.

I've pushed the changes to resolve those two issues and rebased + squashed my commits.

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,117 @@
+            else
+              throw "Could not extract Go version from go.mod";
+
+          go = pkgs."go_${builtins.replaceStrings [ "." ] [ "_" ] goVersion}";
```
I would move this into an overlay to define go to use the exact version of go. If comment is true

Eveeifyeve commented on flake.nix:
```diff
@@ -0,0 +1,117 @@
+            buildInputs = with pkgs; [
+              # Go toolchain
+              go
+              gotools
```
Does this get the version of go that is packaged via nixpkgs or would this use the local version of go defined above?

Eveeifyeve replied:
Yes, but the gotools I know go tools defines a specific version of go inside of its package

doprz replied:
What do you suggest?

Eveeifyeve replied:
> What do you suggest?

https://github.com/jesseduffield/lazygit/pull/4826#discussion_r2365817242

doprz replied:
I've also double checked this by running:
```sh
nix derivation show .#devShells.<system>.default
```
and checking `inputDrvs`.

or

```sh
nix derivation show .#devShells.<system>.default | jq '.[].inputDrvs | keys[]' | grep -e "go-.*drv"
```

doprz commented:
> I've also double checked this by running:
> 
> ```shell
> nix derivation show .#devShells.<system>.default
> ```
> 
> and checking `inputDrvs`.
> 
> or
> 
> ```shell
> nix derivation show .#devShells.<system>.default | jq '.[].inputDrvs | keys[]' | grep -e "go-.*drv"
> ```


@Eveeifyeve I've addressed all PR comments and feedback and I think that this PR is now ready for a final review before @stefanhaller's review. Thank you once again for all of your help!

stefanhaller commented:
I rebased on master, cleaned up the commit message of all the squash junk, and added one more fixup commit for the readme.

@Eveeifyeve Is this good to go from your side? I see that @doprz asked you for one final review, but I didn't see any response to that.

stefanhaller commented:
Since everything seems addressed now, I'm going to merge this. Thanks again @doprz for the contribution, and @Eveeifyeve for the reviews.

stefanhaller commented:
@doprz @Eveeifyeve Can one of you please have a look? We just recently bumped our go requirement to 1.25. I was assuming that this `goVersion` overlay thing would take care of this?

doprz commented:
Thank you once again @Eveeifyeve and @stefanhaller for the reviews and feedback!
